Closes #7

ElvishJerricco opened the pull request:
Closes #7 - Deprecated various initializers and dispatch_object_t properties.

The list of classes below have all been modified in the same way. Each of them had one initializer init(raw:), and initialized the class by assigning its dispatch_object_t subtype property to raw. These initializers have been renamed to init(rawValue:). This is because DispatchObject now requires a get-able property named rawValue whose type is a subtype of dispatch_object_t. These classes each had a property which represented what is now called rawValue. The old properties and initializers have been deprecated in favor of the new ones.

Other initializers have also been modified to be failable (implicitly unwrapped), due to the fact that they would call dispatch_create_X functions which returned implicitly unwrapped optionals. Having failable initializers allows for the ease of use that is expected when these functions are expected not to fail, along with the added ability to handle failure should it occur.
- DispatchData
- DispatchGroup
- DispatchIO
- DispatchQueue
- DispatchSemaphore
- DispatchSource

ElvishJerricco committed:
Closes #7 - Deprecated various initializers and dispatch_object_t properties.

The list of classes below have all been modified in the same way. Each of them had one initializer init(raw:), and initialized the class by assigning its dispatch_object_t subtype property to raw. These initializers have been renamed to init(rawValue:). This is because DispatchObject now requires a get-able property named rawValue whose type is a subtype of dispatch_object_t. These classes each had a property which represented what is now called rawValue. The old properties and initializers have been deprecated in favor of the new ones.

Other initializers have also been modified to be failable (implicitly unwrapped), due to the fact that they would call dispatch_create_X functions which returned implicitly unwrapped optionals. Having failable initializers allows for the ease of use that is expected when these functions are expected not to fail, along with the added ability to handle failure should it occur.

* DispatchData
* DispatchGroup
* DispatchIO
* DispatchQueue
* DispatchSemaphore
* DispatchSource

anpol committed:
strip trailing whitespace

anpol commented:
Please, provide a commit message [describing the essence of the your work](https://github.com/anpol/DispatchKit/blob/master/CONTRIBUTING.rst), something like 'converting from optional data members to failable initializers'.

anpol commented:
Please, run `Tools/strip-trailing-whitespace` script from the project root directory to strip trailing whitespace from this pull-request. I've modified [CONTRIBUTING](https://github.com/anpol/DispatchKit/blob/master/CONTRIBUTING.rst) guide with a note about Xcode settings.

anpol commented on Sources/DispatchObject.swift:
```diff
@@ -8,7 +8,8 @@
 import Foundation
 
 public protocol DispatchObject {
```
I suspect it should extend from `RawRepresentable` protocol.

anpol commented on Sources/DispatchQueue.swift:
```diff
@@ -9,17 +9,25 @@ import Foundation
-
-    public var rawValue: dispatch_object_t! {
-        return queue
+    public typealias RawValue = dispatch_queue_t
```
This `typealias` is redundant, it gets inferred automatically when the `DispatchObject` protocol is adopted. Introducing it here declares an additional nested type `DispatchQueue.RawValue` which is undesirable.
Please, declare `let rawValue: dispatch_queue_t` and remove `RawValue` declaration at the struct level.

anpol commented on Sources/DispatchQueue.swift:
```diff
@@ -9,17 +9,25 @@ import Foundation
-        self.queue = queue
+    public let rawValue: RawValue
+
+    @available(*, unavailable, renamed="DispatchQueue(rawValue:)")
```
If possible, group unavailable definitions together, before `let rawValue`. It quickly becomes hard to read otherwise.

anpol commented on Sources/DispatchData.swift:
```diff
@@ -30,11 +37,16 @@ public struct DispatchData<T: IntegerType>: DispatchObject {
         let size = Scale.toBytes(array.count)
-        self.data = array.withUnsafeBufferPointer { (p) in
+        
+        guard let rawValue = array.withUnsafeBufferPointer({ (p) in
```
(p) → p

anpol commented on Sources/DispatchData.swift:
```diff
@@ -102,6 +136,10 @@ public struct DispatchData<T: IntegerType>: DispatchObject {
 
-public func + <T>(a: DispatchData<T>, b: DispatchData<T>) -> DispatchData<T> {
-    return DispatchData<T>(raw: dispatch_data_create_concat(a.data, b.data))
+public func + <T>(a: DispatchData<T>, b: DispatchData<T>) -> DispatchData<T>? {
```
`DispatchData<T>!` looks like a reasonable return type.

ElvishJerricco committed:
Trimmed whitespace.

ElvishJerricco commented:
I guess I rushed this a little. My bad.

ElvishJerricco committed:
DispatchObject conforms to RawRepresentable

This informed me of a few initializers I'd forgotten about, so those
were added.

ElvishJerricco committed:
Use inferred RawValue type

Don't use the generically named RawValue type. Instead, allow the
compiler to infer what RawValue is, and use the explicit type.

ElvishJerricco committed:
Move "unavailable" annotations to more readable places

Put unavailable things near the top, above their renamed counterparts.

ElvishJerricco committed:
Fix style of single closure parameter in DispatchData.swift

ElvishJerricco committed:
Change DispatchData addition result to implicitly unwrapped

Something would be seriously wrong if it returned nil, so crashing in
that instance is likely preferred. Otherwise, it will be useful to not
have to unwrap.

ElvishJerricco commented:
Alright with everything fixed, the commit message changed, and unit tests passing, do you see anything else that needs changing?

anpol commented:
Committed with minor corrections. Thanks!
